File: backend/gatorsched_api/schemas/manager/requests/manager_requests.py
from pydantic import BaseModel

from gatorsched_api.models.types import EmployeeRequestStatus, ManagerRequestStatus
from gatorsched_api.schemas.employee.requests.employee_requests import (
    RequestPerson,
    RequestShiftSummary,
    RequestType,
)


class ManagerRequestCardBase(BaseModel):
    id: str
    employee_status: EmployeeRequestStatus
    manager_status: ManagerRequestStatus


class ManagerSwapRequestCard(ManagerRequestCardBase):
    type: RequestType = "swap"

    requester: RequestPerson
    coverEmployee: RequestPerson

    requesterShift: RequestShiftSummary
    coverShift: RequestShiftSummary

    impactSummary: str | None = None


ManagerRequestCard = ManagerSwapRequestCard


class ManagerRequestCardGroup(BaseModel):
    role: str
    requests: list[ManagerRequestCard]


class ManagerRequestCardResponse(BaseModel):
    groups: list[ManagerRequestCardGroup]

File: frontend/src/services/requests/managerRequests.ts
import { API_BASE } from '../apiConfig';
import {
  EmployeeRequestStatus,
  ManagerRequestStatus,
  RequestPerson,
  RequestShiftSummary,
} from './types';

export type ManagerRequestCardBase = {
  id: string;
  employee_status: EmployeeRequestStatus;
  manager_status: ManagerRequestStatus;
};

export type ManagerSwapRequestCard = ManagerRequestCardBase & {
  type: 'swap';
  requester: RequestPerson;
  coverEmployee: RequestPerson;
  requesterShift: RequestShiftSummary;
  coverShift: RequestShiftSummary;
  impactSummary: string | null;
};

export type ManagerRequestCard = ManagerSwapRequestCard;

export type ManagerRequestCardGroup = {
  role: string;
  requests: ManagerRequestCard[];
};

export async function getManagerRequests(): Promise<ManagerRequestCardGroup[]> {
  const res = await fetch(`${API_BASE}/manager/requests`);

  if (!res.ok) {
    throw new Error('Failed to retrieve Manager Requests');
  }

  const data = await res.json();

  return data.groups;
}
